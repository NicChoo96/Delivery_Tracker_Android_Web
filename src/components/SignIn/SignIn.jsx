import React, { useState, useRef, useEffect } from 'react';
import './SignIn.scss';
import { makeStyles } from '@material-ui/core/styles';
import {
  Button,
  TextField,
  Link,
  Grid,
  Box,
  Container,
  Typography,
} from '@material-ui/core';
import { Alert } from '@material-ui/lab';
import { withFirebase } from '../../Firebase';
import { history } from '../../utils';
import { loginFailureEasterEgg } from '../../constants';

const useStyles = makeStyles((theme) => ({
  paper: {
    marginTop: theme.spacing(8),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
  },
  form: {
    width: '100%',
    marginTop: theme.spacing(1),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
}));


/**
 * This component will render the SignIn component for existing users to input their email and password
 * @category UI Component
 * @subcategory SignIn
 * @Class SignIn
 * @param {Firebase} firebase CRUD interface to communiate with firebase backend service
 * @param {function} toggleShowSignIn toggles which to show between SignIn and SignUp
 * @return {ReactComponent}
 */
function SignIn({ firebase, toggleShowSignIn }) {
  const classes = useStyles();
  const [errorMessage, setErrorMessage] = useState(null);
  const [signinFailureCounter, setSigninFailureCounter] = useState(0);
  const [signinFailureMessages] = useState(loginFailureEasterEgg);
  const [brandingMessage, setBrandingMessage] = useState(
    'Please login or signup for an account.'
  );

  const usernameInputRef = useRef(null);
  const passwordInputRef = useRef(null);

  useEffect(() => {
    if (firebase === null) return;

    firebase.auth.onAuthStateChanged((userAuth) => {
      if (userAuth) {
        history.push('/jobs');
      }
    });
  }, [firebase]);

  const handleChange = (event) => {
    event.preventDefault();
    // do nothing
  };

  const signUpHandler = (event) => {
    event.preventDefault();
    event.stopPropagation();

    // change to signup component
    toggleShowSignIn(false);
  };

  const handleSubmit = (event) => {
    event.preventDefault();
    firebase
      .signIn(usernameInputRef.current.value, passwordInputRef.current.value)
      .then(() => {
        // routing away from the signin page when success
        setErrorMessage(null);
        history.push('/jobs');
      })
      .catch((exception) => {
        console.error(exception);

        // setting easter egs
        setBrandingMessage(
          signinFailureMessages[
          signinFailureCounter % signinFailureMessages.length
          ]
        );
        setSigninFailureCounter((state) => state + 1);

        // displaying error alert based on error received from firebase
        switch (exception.code) {
          case 'auth/invalid-email':
            setErrorMessage('Invalid email');
            break;
          case 'auth/user-not-found':
            setErrorMessage('No such user exist');
            break;
          default:
            setErrorMessage('Invalid username and password combination');
            break;
        }
      });
  };

  return (
    <Container maxWidth="md">
      <Grid container justify="center" alignItems="center">
        <Grid item md={5}>
          <div className="signin__branding">
            <span component="p" className="signin__branding-header">
              Trackee Co.
            </span>
            <span component="p" className="signin__branding-message">
              {brandingMessage}
            </span>
          </div>

          <form onSubmit={handleSubmit} onChange={handleChange} noValidate>
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              id="signin_username"
              label="Email"
              name="Email"
              autoComplete="email"
              autoFocus
              inputRef={usernameInputRef}
            />

            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              name="password"
              label="Password"
              type="password"
              id="signin_password"
              autoComplete="current-password"
              inputRef={passwordInputRef}
            />
            <span className="signin__submit-btn-group">
              <Button
                type="submit"
                variant="contained"
                color="primary"
                className={classes.submit}
              >
                Sign In
              </Button>
              <Button
                onClick={signUpHandler}
                variant="outlined"
                color="primary"
                className={classes.submit}
              >
                Sign Up
              </Button>
            </span>
            {errorMessage != null ? (
              <Alert severity="error">{errorMessage}</Alert>
            ) : (
                <span></span>
              )}
          </form>
          <Box mt={8}>
            <Typography variant="body2" color="textSecondary" align="center">
              {'Copyright Â© '}
              <Link color="inherit" href="https://material-ui.com/">
                CZ3003 - SSP1 - Staizen
              </Link>
            </Typography>
          </Box>
        </Grid>
      </Grid>
    </Container>
  );
}

export default withFirebase(SignIn);
