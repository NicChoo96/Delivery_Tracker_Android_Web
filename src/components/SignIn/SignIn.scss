.signin__submit-btn-group {
  display: flex;
  justify-content: space-between;
}

.signin__login-img {
  width: auto;
  height:100%
}

.signin__branding {
  display: flex;
  flex-direction: column;
  text-align: center;
  user-select: none;
}

.signin__branding-header {
  font-size: 3em;
  font-weight: 900;
}

.signin__branding-message {
  color: grey;
  font-style: italic;
  height: 4em;
}
