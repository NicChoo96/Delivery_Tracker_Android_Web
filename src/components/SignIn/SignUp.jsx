import React, { useState, useRef } from 'react';
import './SignIn.scss';
import { makeStyles } from '@material-ui/core/styles';
import {
  Button,
  TextField,
  Link,
  Grid,
  Box,
  Container,
  Typography,
} from '@material-ui/core';
import { Alert } from '@material-ui/lab';
import { withFirebase } from '../../Firebase';

const useStyles = makeStyles((theme) => ({
  paper: {
    marginTop: theme.spacing(8),
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
  },
  avatar: {
    margin: theme.spacing(1),
    backgroundColor: theme.palette.secondary.main,
  },
  form: {
    width: '100%',
    marginTop: theme.spacing(1),
  },
  submit: {
    margin: theme.spacing(3, 0, 2),
  },
}));


/**
 * This component will render the SignIn component for existing users to input their email and password
 * @category UI Component
 * @subcategory SignIn
 * @Class SignUp
 * @param {Firebase} firebase CRUD interface to communiate with firebase backend service
 * @param {function} toggleShowSignIn toggles which to show between SignIn and SignUp
 * @return {ReactComponent}
 */
function SignUp({ firebase, toggleShowSignIn }) {
  const classes = useStyles();
  const [errorMessage, setErrorMessage] = useState(null);

  const usernameInputRef = useRef(null);
  const passwordInputRef = useRef(null);
  const confirmPasswordInputRef = useRef(null);

  const handleChange = (event) => {
    event.preventDefault();
    // do nothing
  };

  const backToSignIn = (event) => {
    event.preventDefault();
    event.stopPropagation();

    // change back to signin component
    toggleShowSignIn(true);
  };

  const handleSubmit = (event) => {
    event.preventDefault();

    const PasswordMatch = passwordInputRef.current.value === confirmPasswordInputRef.current.value;

    setErrorMessage(PasswordMatch ? null : 'Password Do Not Match');

    // storing data in firebase and email validation
    firebase.signUp(
      usernameInputRef.current.value,
      passwordInputRef.current.value
    ).then(() => {
      // routing away from the signup page when success
      setErrorMessage(null);
    }).catch((exception) => {
      console.error(exception)
      switch (exception.code) {
        case "auth/weak-password":
          setErrorMessage('Password should be at least 6 characters');
          break;
        case "auth/invalid-email":
          setErrorMessage('Invalid Email');
          break;
        case "auth/email-already-in-use":
          setErrorMessage('Email already in used');
          break;
        default:
          setErrorMessage(`${exception.code} : ${exception.message}`);
          break;
      }
    });
  };

  return (
    <Container maxWidth="md">
      <Grid container justify="center" alignItems="center">
        <Grid item md={5}>
          <div className="signin__branding">
            <span component="p" className="signin__branding-header">
              Trackee Co.
            </span>
            <span component="p" className="signin__branding-message">
              {`It's quick and easy.`}
            </span>
          </div>

          <form onSubmit={handleSubmit} onChange={handleChange} noValidate>
            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              id="signup_username"
              label="Email"
              name="Email"
              type="email"
              autoFocus
              inputRef={usernameInputRef}
            />

            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              name="password"
              label="Password"
              type="password"
              id="signup_password"
              inputRef={passwordInputRef}
            />

            <TextField
              variant="outlined"
              margin="normal"
              required
              fullWidth
              name="confirm password"
              label="Confirm Password"
              type="password"
              id="signup_password_confirmation"
              inputRef={confirmPasswordInputRef}
            />

            <span className="signin__submit-btn-group">
              <Button
                onClick={backToSignIn}
                variant="outlined"
                color="primary"
                className={classes.submit}
              >
                back to Sign In
              </Button>
              <Button
                type="submit"
                variant="contained"
                color="primary"
                className={classes.submit}
              >
                Sign Up
              </Button>
            </span>
            {errorMessage != null ? (
              <Alert severity="error">{errorMessage}</Alert>
            ) : (
                <span></span>
              )}
          </form>
          <Box mt={8}>
            <Typography variant="body2" color="textSecondary" align="center">
              {'Copyright Â© '}
              <Link color="inherit" href="https://material-ui.com/">
                CZ3003 - SSP1 - Staizen
              </Link>
            </Typography>
          </Box>
        </Grid>
      </Grid>
    </Container>
  );
}

export default withFirebase(SignUp);
