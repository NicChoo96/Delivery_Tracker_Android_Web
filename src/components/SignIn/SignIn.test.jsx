import React from 'react';
import ReactDOM from 'react-dom';
import '@testing-library/jest-dom/extend-expect';

import { cleanup } from '@testing-library/react';
import renderer from 'react-test-renderer';

import SignIn from './SignIn';

afterEach(cleanup);

describe('SignIn', () => {
  test('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<SignIn />, div);
  });

  // we do not have a prop for this component, so the second test is omitted

  test('matches snapshot', () => {
    const tree = renderer.create(<SignIn />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
