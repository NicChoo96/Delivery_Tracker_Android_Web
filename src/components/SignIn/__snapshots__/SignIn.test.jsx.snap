// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn matches snapshot 1`] = `
<div
  className="MuiContainer-root MuiContainer-maxWidthMd"
>
  <div
    className="MuiGrid-root MuiGrid-container MuiGrid-align-items-xs-center MuiGrid-justify-xs-center"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-md-5"
    >
      <div
        className="signin__branding"
      >
        <span
          className="signin__branding-header"
          component="p"
        >
          Trackee Co.
        </span>
        <span
          className="signin__branding-message"
          component="p"
        >
          Please login or signup for an account.
        </span>
      </div>
      <form
        noValidate={true}
        onChange={[Function]}
        onSubmit={[Function]}
      >
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
            data-shrink={false}
            htmlFor="signin_username"
            id="signin_username-label"
          >
            Email
            <span
              aria-hidden={true}
              className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoComplete="email"
              autoFocus={true}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              id="signin_username"
              name="Email"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="text"
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-7"
              >
                <span>
                  Email
                   *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          className="MuiFormControl-root MuiTextField-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined Mui-required Mui-required"
            data-shrink={false}
            htmlFor="signin_password"
            id="signin_password-label"
          >
            Password
            <span
              aria-hidden={true}
              className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
            >
               
              *
            </span>
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-fullWidth MuiInputBase-formControl"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoComplete="current-password"
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input"
              disabled={false}
              id="signin_password"
              name="password"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={true}
              type="password"
            />
            <fieldset
              aria-hidden={true}
              className="PrivateNotchedOutline-root-5 MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="PrivateNotchedOutline-legendLabelled-7"
              >
                <span>
                  Password
                   *
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <span
          className="signin__submit-btn-group"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-contained makeStyles-submit-4 MuiButton-containedPrimary"
            disabled={false}
            onBlur={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="submit"
          >
            <span
              className="MuiButton-label"
            >
              Sign In
            </span>
          </button>
          <button
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined makeStyles-submit-4 MuiButton-outlinedPrimary"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiButton-label"
            >
              Sign Up
            </span>
          </button>
        </span>
        <span />
      </form>
      <div
        className="MuiBox-root MuiBox-root-10"
      >
        <p
          className="MuiTypography-root MuiTypography-body2 MuiTypography-colorTextSecondary MuiTypography-alignCenter"
        >
          Copyright © 
          <a
            className="MuiTypography-root MuiLink-root MuiLink-underlineHover MuiTypography-colorInherit"
            href="https://material-ui.com/"
            onBlur={[Function]}
            onFocus={[Function]}
          >
            CZ3003 - SSP1 - Staizen
          </a>
        </p>
      </div>
    </div>
  </div>
</div>
`;
